crates/apollo-router/src/http_server_factory.rs

```diff
@@ -112,27 +120,103 @@ impl HttpServerHandle {
         Ok(handle)
     }
 
-    pub(crate) fn listen_address(&self) -> SocketAddr {
-        self.listen_address
+    pub(crate) fn listen_address(&self) -> &ListenAddr {
+        &self.listen_address
+    }
+}
+
+pub(crate) trait Listener: Send + Unpin {
+    fn accept(&self) -> future::BoxFuture<io::Result<(BoxAsyncReadWrite, ListenAddr)>>;
+    fn local_addr(&self) -> io::Result<ListenAddr>;
+}
+
+impl Listener for TcpListener {
+    fn accept(&self) -> future::BoxFuture<io::Result<(BoxAsyncReadWrite, ListenAddr)>> {
+        self.accept()
+            .map(|res| {
+                let (stream, addr) = res?;
+                Ok((Box::new(stream) as Box<_>, addr.into()))
+            })
+            .boxed()
+    }
+    fn local_addr(&self) -> io::Result<ListenAddr> {
+        self.local_addr().map(Into::into)
+    }
+}
+
+#[cfg(unix)]
+impl Listener for UnixListener {
+    fn accept(&self) -> future::BoxFuture<io::Result<(BoxAsyncReadWrite, ListenAddr)>> {
+        self.accept()
+            .map(|res| {
+                let (stream, addr) = res?;
+                Ok((Box::new(stream) as Box<_>, addr.into()))
+            })
+            .boxed()
+    }
+    fn local_addr(&self) -> io::Result<ListenAddr> {
+        self.local_addr().map(Into::into)
+    }
+}
+
+pub(crate) trait AsyncReadWrite: tokio::io::AsyncRead + tokio::io::AsyncWrite {
+    fn set_nodelay(&self, nodelay: bool) -> io::Result<()>;
+}
+
+pub(crate) type BoxAsyncReadWrite = Box<dyn AsyncReadWrite + Send + Unpin + 'static>;
```

Comment:
it's not a good idea to have a `Box<dyn>` to hold either a `TcpStream` or `UnixStream`. It is useful when we're handling both kinds of objects at runtime, but here we're always creating one kind of object (we won't run in TCP and Unix socket modes at the same time), so we should not bear the cost of adding an allocation in the hot path for that.
`AsyncRead` has one method, `AsyncWrite` has 5 methods, so it's still reasonable to make an enum here.

Reply:
Fixed with new impl